### PR TITLE
Add additional `Compiler`/`Context` functions for getting compiler configuration and main func/output types

### PR DESCRIPTION
**Pull request by SomeRanDev**
At the moment, it's impossible to get the compiler arguments and learn what the main function is with Haxe macros. This pull seeks to resolve this by exposing additional information from the context object in OCaml.

### Input
There's a collection of static info regarding compiler input that could be helpful in Haxe macros, such as compiler arguments, main class path, platform, version, etc. The `Compiler.getConfiguration` function helps provide this info by exposing the "config" fields of the [OCaml context object](https://github.com/HaxeFoundation/haxe/blob/13a8e48cbc5a58ffd77647e7d2107ecafec604d8/src/context/common.ml#L339).

### Output
Similarly, there is some output information within this context object that would also be nice to access from `Context`. At the current moment, only the JavaScript target can access the generated "main" typed expression and a list of `haxe.macro.Type`s  outputs using `setCustomJSGenerator` and `JSGenApi`.

The `Context.getMainExpr()` allows access to the main expression from any target at compile-time.

`Context.getAllModuleTypes()` on the other hand provides a list of the `haxe.macro.ModuleType`s that are readily available after a certain point in the compilation process. In order to keep consistent with the capabilities of the JS target, the two OCaml functions used to converting between `ModuleType`s and `Type`s have also been exposed in `haxe.macro.TypeTools`.

---

### Functions added
* `haxe.macro.Compiler.getConfiguration`
* `haxe.macro.Context.getMainExpr`
* `haxe.macro.Context.getAllModuleTypes`
* `haxe.macro.TypeTools.toModuleType`
* `haxe.macro.TypeTools.fromModuleType`

Any thoughts are appreciated!

### CODE REVIEW

**Comment by Simn** [comment]
Maybe we should have something like `Compiler.getConfiguration` which returns an object, so that we don't need a new function for everything...

Also, I don't see why `getAllTypes` should exist, it conveys the same information as `getAllModuleTypes`, just in a different format.

**Comment by SomeRanDev** [comment]
I'd absolutely love `getConfiguration` especially if it means everything not already visible from the [(* config *)](https://github.com/HaxeFoundation/haxe/blob/13a8e48cbc5a58ffd77647e7d2107ecafec604d8/src/context/common.ml#L339) section can be added. 

And yeahhh, originally just had `getAllModuleTypes`, but noticed the JSGen system converted and used Types, so copied and threw in that one because there's no function for (easily) converting between the two on the Haxe-side. But yeah, definitely would be better to remove that and instead expose or implement a Haxe version of the [type_of_module_type](https://github.com/HaxeFoundation/haxe/blob/31779014a0e4edbb0acf3c999cb5ba694c26dc3c/src/core/tFunctions.ml#L645) func (and vise versa) in `haxe.macro.TypeTools`, so lemme try that.

File: src/macro/macroApi.ml [hunk 1]
```diff
@@ -403,6 +414,97 @@ and encode_display_mode dm =
 	in
 	encode_enum ~pos:None IDisplayMode tag pl
 
+(** encoded to haxe.display.Display.Platform, an enum abstract of String *)
+and encode_platform p =
```
Comment (kLabz) [inline src/macro/macroApi.ml:420]: You can use existing Globals.platform_name

**Comment by Simn** [comment]
Oh wow, you went all-in on that configuration part!

My only concern in that regard is that we now can't easily change these internal structures without also changing the public interface. I suppose we could add a documentation comment about that somewhere so that people know something _might_ change between compiler versions. It's a rare occurrence anyway, but I want to keep the option open at least.

File: src/macro/macroApi.ml (hunk 1 above)
Comment (SomeRanDev) [reply src/macro/macroApi.ml:420]: Oh, nice catch! thank you! 😄

**Comment by SomeRanDev** [comment]
Haha, just wanted to see what it would look like if everything was typed, then get some feedback and trim down from there. (;^_^) `PlatformConfig` might be a bit overkill, but definitely some things like checking for non-nullable primitive types and allowing target overload functions can be helpful in Haxe macros. 

Alternatively, since most of the `PlatformConfig` fields are Bool, could perhaps make it a `Map<String,Bool>` (with some of the enums translated to Bool like `VarScope` translating to `isScopeFunctionBased` or `isScopeBlockBased`). Of course, having a typed system would be nicer, but a map would cut down a majority of the public api maintenance while still keeping 99% of the usability (or maybe it's worse since you'd still need to keep an updated list of possible map keys).

**Comment by SomeRanDev** [comment]
Mostly copied the warning comment from [haxe.macro.Type](https://github.com/HaxeFoundation/haxe/blob/bf9ff69c0801082174f0b2b0a66faeb5356de580/std/haxe/macro/Type.hx#L25), seemed very applicable.

Let me know if there's anything else, and I'll work on adding tests in meantime! 👍

**Comment by Simn** [comment]
Nice!

There might be some minor inaccuracies in the documentation, but we can address that later. Thank you for the contribution!

**Comment by kLabz** [comment]
Ha, had a minor report in the documentation indeed but was about to say lgtm but you just merged it :D

**Comment by SomeRanDev** [comment]
Oh wow, that was fast! Thank you all so much, you have no idea how much these features will help! 😁❤️

(_Sorry about the docs, ooof._)

**Pull request merged**